File: src/hbse/core/policy.py
"""Deny-by-default policy evaluation."""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import UTC, datetime
from enum import StrEnum

from pydantic import BaseModel, ConfigDict, Field


class DeliveryMode(StrEnum):
    BROKERED_HTTP = "brokered_http"
    BROKERED_OPERATION = "brokered_operation"
    CALLBACK = "callback"
    PIPE = "pipe"
    FD = "fd"
    TEMP_FILE = "temp_file"
    CHILD_ENV = "child_env"
    RAW = "raw"
    TERMINAL_PRINT = "terminal_print"


class PolicyDecision(StrEnum):
    ALLOW = "allow"
    DENY = "deny"


class AccessPolicy(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    policy_id: str
    secret_refs: list[str] = Field(default_factory=list)
    allowed_consumers: list[str] = Field(default_factory=list)
    denied_consumers: list[str] = Field(default_factory=list)
    allowed_purposes: list[str] = Field(default_factory=list)
    denied_purposes: list[str] = Field(default_factory=list)
    allowed_delivery_modes: list[DeliveryMode] = Field(default_factory=list)
    allowed_http_hosts: list[str] = Field(default_factory=list)
    denied_http_hosts: list[str] = Field(default_factory=list)
    allowed_http_methods: list[str] = Field(default_factory=list)
    denied_http_methods: list[str] = Field(default_factory=list)
    allowed_http_path_prefixes: list[str] = Field(default_factory=list)
    denied_http_path_prefixes: list[str] = Field(default_factory=list)
    require_https_for_brokered_http: bool = False
    max_http_request_body_bytes: int | None = Field(default=None, ge=0)
    allowed_os_uids: list[int] = Field(default_factory=list)
    denied_os_uids: list[int] = Field(default_factory=list)
    allowed_executable_paths: list[str] = Field(default_factory=list)
    denied_executable_paths: list[str] = Field(default_factory=list)
    allowed_executable_sha256: list[str] = Field(default_factory=list)
    denied_executable_sha256: list[str] = Field(default_factory=list)
    exportable: bool = False
    max_ticket_ttl_seconds: int = Field(default=60, ge=1, le=3600)
    max_uses: int = Field(default=1, ge=1, le=100)
    minimum_provider_assurance: str = "A1"
    expires_at: datetime | None = None


@dataclass(frozen=True)
class AccessRequest:
    secret_ref: str
    consumer: str
    purpose: str
    delivery_mode: DeliveryMode
    provider_assurance: str
    raw_export_requested: bool = False
    http_host: str | None = None
    http_scheme: str | None = None
    http_method: str | None = None
    http_path: str | None = None
    http_request_body_bytes: int | None = None
    os_uid: int | None = None
    executable_path: str | None = None
    executable_sha256: str | None = None
    now: datetime = field(default_factory=lambda: datetime.now(UTC))


@dataclass(frozen=True)
class EvaluationResult:
    decision: PolicyDecision
    reason: str
    policy: AccessPolicy | None = None

    @property
    def allowed(self) -> bool:
        return self.decision == PolicyDecision.ALLOW


class PolicyEngine:
    def __init__(self, policies: list[AccessPolicy] | None = None) -> None:
        self._policies = policies or []

    def evaluate(self, request: AccessRequest) -> EvaluationResult:
        for policy in self._policies:
            result = self._evaluate_policy(policy, request)
            if result.decision == PolicyDecision.ALLOW:
                return result
        return EvaluationResult(PolicyDecision.DENY, "no matching allow policy")

    def _evaluate_policy(self, policy: AccessPolicy, request: AccessRequest) -> EvaluationResult:
        if request.secret_ref not in policy.secret_refs and "*" not in policy.secret_refs:
            return EvaluationResult(PolicyDecision.DENY, "secret reference not covered", policy)
        if policy.expires_at and request.now > policy.expires_at.astimezone(UTC):
            return EvaluationResult(PolicyDecision.DENY, "policy expired", policy)
        if request.consumer in policy.denied_consumers:
            return EvaluationResult(PolicyDecision.DENY, "consumer explicitly denied", policy)
        if request.purpose in policy.denied_purposes:
            return EvaluationResult(PolicyDecision.DENY, "purpose explicitly denied", policy)
        if request.consumer not in policy.allowed_consumers and "*" not in policy.allowed_consumers:
            return EvaluationResult(PolicyDecision.DENY, "consumer not allowed", policy)
        if request.purpose not in policy.allowed_purposes and "*" not in policy.allowed_purposes:
            return EvaluationResult(PolicyDecision.DENY, "purpose not allowed", policy)
        if request.delivery_mode not in policy.allowed_delivery_modes:
            return EvaluationResult(PolicyDecision.DENY, "delivery mode not allowed", policy)
        if request.os_uid in policy.denied_os_uids:
            return EvaluationResult(PolicyDecision.DENY, "OS user explicitly denied", policy)
        if policy.allowed_os_uids and request.os_uid not in policy.allowed_os_uids:
            return EvaluationResult(PolicyDecision.DENY, "OS user not allowed", policy)
        if request.executable_path in policy.denied_executable_paths:
            return EvaluationResult(PolicyDecision.DENY, "executable path explicitly denied", policy)
        if policy.allowed_executable_paths and request.executable_path not in policy.allowed_executable_paths:
            return EvaluationResult(PolicyDecision.DENY, "executable path not allowed", policy)
        if request.executable_sha256 in policy.denied_executable_sha256:
            return EvaluationResult(PolicyDecision.DENY, "executable hash explicitly denied", policy)
        if policy.allowed_executable_sha256 and request.executable_sha256 not in policy.allowed_executable_sha256:
            return EvaluationResult(PolicyDecision.DENY, "executable hash not allowed", policy)
        if request.delivery_mode == DeliveryMode.BROKERED_HTTP:
            if request.http_host is None:
                return EvaluationResult(PolicyDecision.DENY, "HTTP host required for brokered_http", policy)
            if policy.require_https_for_brokered_http and request.http_scheme != "https":
                return EvaluationResult(PolicyDecision.DENY, "HTTPS required for brokered_http", policy)
            if request.http_host in policy.denied_http_hosts:
                return EvaluationResult(PolicyDecision.DENY, "HTTP host explicitly denied", policy)
            if policy.allowed_http_hosts and request.http_host not in policy.allowed_http_hosts and "*" not in policy.allowed_http_hosts:
                return EvaluationResult(PolicyDecision.DENY, "HTTP host not allowed", policy)
            if request.http_method is None:
                return EvaluationResult(PolicyDecision.DENY, "HTTP method required for brokered_http", policy)
            method = request.http_method.upper()
            allowed_methods = [item.upper() for item in policy.allowed_http_methods]
            denied_methods = [item.upper() for item in policy.denied_http_methods]
            if method in denied_methods:
                return EvaluationResult(PolicyDecision.DENY, "HTTP method explicitly denied", policy)
            if allowed_methods and method not in allowed_methods and "*" not in allowed_methods:
                return EvaluationResult(PolicyDecision.DENY, "HTTP method not allowed", policy)
            if request.http_path is None:
                return EvaluationResult(PolicyDecision.DENY, "HTTP path required for brokered_http", policy)
            if any(request.http_path.startswith(prefix) for prefix in policy.denied_http_path_prefixes):
                return EvaluationResult(PolicyDecision.DENY, "HTTP path explicitly denied", policy)
            if policy.allowed_http_path_prefixes and not any(
                request.http_path.startswith(prefix) for prefix in policy.allowed_http_path_prefixes
            ):
                return EvaluationResult(PolicyDecision.DENY, "HTTP path not allowed", policy)
            if (
                policy.max_http_request_body_bytes is not None
                and request.http_request_body_bytes is not None
                and request.http_request_body_bytes > policy.max_http_request_body_bytes
            ):
                return EvaluationResult(PolicyDecision.DENY, "HTTP request body too large", policy)
        if request.raw_export_requested and not policy.exportable:
            return EvaluationResult(PolicyDecision.DENY, "raw export not allowed", policy)
        if _assurance_rank(request.provider_assurance) < _assurance_rank(policy.minimum_provider_assurance):
            return EvaluationResult(PolicyDecision.DENY, "provider assurance too low", policy)
        return EvaluationResult(PolicyDecision.ALLOW, "allowed", policy)


def _assurance_rank(level: str) -> int:
    order = {"A0": 0, "A1": 1, "A2": 2, "A3": 3, "A4": 4, "A5": 5}
    return order.get(level, -1)

File: hbse/src/policy.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum DeliveryMode {
    BrokeredHttp,
    BrokeredOperation,
    Callback,
    Pipe,
    Fd,
    TempFile,
    ChildEnv,
    Raw,
    TerminalPrint,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct AccessPolicy {
    pub policy_id: String,
    #[serde(default)]
    pub secret_refs: Vec<String>,
    #[serde(default)]
    pub allowed_consumers: Vec<String>,
    #[serde(default)]
    pub denied_consumers: Vec<String>,
    #[serde(default)]
    pub allowed_purposes: Vec<String>,
    #[serde(default)]
    pub denied_purposes: Vec<String>,
    #[serde(default)]
    pub allowed_delivery_modes: Vec<DeliveryMode>,
    #[serde(default)]
    pub allowed_http_hosts: Vec<String>,
    #[serde(default)]
    pub denied_http_hosts: Vec<String>,
    #[serde(default)]
    pub allowed_http_methods: Vec<String>,
    #[serde(default)]
    pub denied_http_methods: Vec<String>,
    #[serde(default)]
    pub allowed_http_path_prefixes: Vec<String>,
    #[serde(default)]
    pub denied_http_path_prefixes: Vec<String>,
    #[serde(default)]
    pub require_https_for_brokered_http: bool,
    #[serde(default)]
    pub max_http_request_body_bytes: Option<u64>,
    #[serde(default)]
    pub allowed_os_uids: Vec<u32>,
    #[serde(default)]
    pub denied_os_uids: Vec<u32>,
    #[serde(default)]
    pub allowed_executable_paths: Vec<String>,
    #[serde(default)]
    pub denied_executable_paths: Vec<String>,
    #[serde(default)]
    pub allowed_executable_sha256: Vec<String>,
    #[serde(default)]
    pub denied_executable_sha256: Vec<String>,
    #[serde(default)]
    pub exportable: bool,
    #[serde(default = "default_ticket_ttl")]
    pub max_ticket_ttl_seconds: u64,
    #[serde(default = "default_max_uses")]
    pub max_uses: u32,
    #[serde(default = "default_minimum_provider_assurance")]
    pub minimum_provider_assurance: String,
    #[serde(default)]
    pub require_mfa: bool,
    #[serde(default)]
    pub allow_unbound_plaintext_export: bool,
    #[serde(default)]
    pub expires_at: Option<DateTime<Utc>>,
}

fn default_ticket_ttl() -> u64 {
    60
}

fn default_max_uses() -> u32 {
    1
}

fn default_minimum_provider_assurance() -> String {
    "A1".to_string()
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct AccessRequest {
    pub secret_ref: String,
    pub consumer: String,
    pub purpose: String,
    pub delivery_mode: DeliveryMode,
    pub provider_assurance: String,
    pub raw_export_requested: bool,
    pub http_host: Option<String>,
    pub http_scheme: Option<String>,
    pub http_method: Option<String>,
    pub http_path: Option<String>,
    pub http_request_body_bytes: Option<u64>,
    pub os_uid: Option<u32>,
    pub executable_path: Option<String>,
    pub executable_sha256: Option<String>,
    pub mfa_verified: bool,
    pub broker_session_id: Option<String>,
    pub now: DateTime<Utc>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum PolicyDecision {
    Allow,
    Deny,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct EvaluationResult<'a> {
    pub decision: PolicyDecision,
    pub reason: String,
    pub policy: Option<&'a AccessPolicy>,
}

impl EvaluationResult<'_> {
    pub fn allowed(&self) -> bool {
        self.decision == PolicyDecision::Allow
    }
}

#[derive(Debug, Clone)]
pub struct PolicyEngine {
    policies: Vec<AccessPolicy>,
}

impl PolicyEngine {
    pub fn new(policies: Vec<AccessPolicy>) -> Self {
        Self { policies }
    }

    pub fn evaluate(&self, request: &AccessRequest) -> EvaluationResult<'_> {
        for policy in &self.policies {
            let result = self.evaluate_policy(policy, request);
            if result.decision == PolicyDecision::Allow {
                return result;
            }
        }
        EvaluationResult {
            decision: PolicyDecision::Deny,
            reason: "no matching allow policy".to_string(),
            policy: None,
        }
    }

    fn evaluate_policy<'a>(
        &'a self,
        policy: &'a AccessPolicy,
        request: &AccessRequest,
    ) -> EvaluationResult<'a> {
        macro_rules! deny {
            ($reason:expr) => {
                return EvaluationResult {
                    decision: PolicyDecision::Deny,
                    reason: $reason.to_string(),
                    policy: Some(policy),
                }
            };
        }

        if !policy.secret_refs.contains(&request.secret_ref)
            && !policy.secret_refs.iter().any(|item| item == "*")
        {
            deny!("secret reference not covered");
        }
        if let Some(expires_at) = policy.expires_at {
            if request.now > expires_at {
                deny!("policy expired");
            }
        }
        if policy.denied_consumers.contains(&request.consumer) {
            deny!("consumer explicitly denied");
        }
        if policy.denied_purposes.contains(&request.purpose) {
            deny!("purpose explicitly denied");
        }
        if !policy.allowed_consumers.contains(&request.consumer)
            && !policy.allowed_consumers.iter().any(|item| item == "*")
        {
            deny!("consumer not allowed");
        }
        if !policy.allowed_purposes.contains(&request.purpose)
            && !policy.allowed_purposes.iter().any(|item| item == "*")
        {
            deny!("purpose not allowed");
        }
        if !policy
            .allowed_delivery_modes
            .contains(&request.delivery_mode)
        {
            deny!("delivery mode not allowed");
        }
        if let Some(uid) = request.os_uid {
            if policy.denied_os_uids.contains(&uid) {
                deny!("OS user explicitly denied");
            }
        }
        if !policy.allowed_os_uids.is_empty()
            && !request
                .os_uid
                .map_or(false, |uid| policy.allowed_os_uids.contains(&uid))
        {
            deny!("OS user not allowed");
        }
        if let Some(path) = &request.executable_path {
            if policy.denied_executable_paths.contains(path) {
                deny!("executable path explicitly denied");
            }
        }
        if !policy.allowed_executable_paths.is_empty()
            && !request
                .executable_path
                .as_ref()
                .map_or(false, |path| policy.allowed_executable_paths.contains(path))
        {
            deny!("executable path not allowed");
        }
        if let Some(hash) = &request.executable_sha256 {
            if policy.denied_executable_sha256.contains(hash) {
                deny!("executable hash explicitly denied");
            }
        }
        if !policy.allowed_executable_sha256.is_empty()
            && !request.executable_sha256.as_ref().map_or(false, |hash| {
                policy.allowed_executable_sha256.contains(hash)
            })
        {
            deny!("executable hash not allowed");
        }
        if request.delivery_mode == DeliveryMode::BrokeredHttp {
            if let Err(result) = self.evaluate_brokered_http(policy, request) {
                return result;
            }
        }
        if request.raw_export_requested && !policy.exportable {
            deny!("raw export not allowed");
        }
        if request.raw_export_requested
            && matches!(
                request.delivery_mode,
                DeliveryMode::Raw | DeliveryMode::TerminalPrint
            )
            && !policy.allow_unbound_plaintext_export
            && !policy_has_peer_binding(policy)
        {
            deny!("raw plaintext export requires a non-spoofable peer binding");
        }
        if assurance_rank(&request.provider_assurance)
            < assurance_rank(&policy.minimum_provider_assurance)
        {
            deny!("provider assurance too low");
        }
        if policy.require_mfa && !request.mfa_verified {
            deny!("MFA required");
        }
        EvaluationResult {
            decision: PolicyDecision::Allow,
            reason: "allowed".to_string(),
            policy: Some(policy),
        }
    }

    fn evaluate_brokered_http<'a>(
        &self,
        policy: &'a AccessPolicy,
        request: &AccessRequest,
    ) -> Result<(), EvaluationResult<'a>> {
        macro_rules! deny {
            ($reason:expr) => {
                return Err(EvaluationResult {
                    decision: PolicyDecision::Deny,
                    reason: $reason.to_string(),
                    policy: Some(policy),
                })
            };
        }
        let Some(host) = request.http_host.as_ref() else {
            deny!("HTTP host required for brokered_http");
        };
        if policy.require_https_for_brokered_http && request.http_scheme.as_deref() != Some("https")
        {
            deny!("HTTPS required for brokered_http");
        }
        if policy.denied_http_hosts.contains(host) {
            deny!("HTTP host explicitly denied");
        }
        if !policy.allowed_http_hosts.is_empty()
            && !policy.allowed_http_hosts.contains(host)
            && !policy.allowed_http_hosts.iter().any(|item| item == "*")
        {
            deny!("HTTP host not allowed");
        }
        let Some(method) = request.http_method.as_ref() else {
            deny!("HTTP method required for brokered_http");
        };
        let method = method.to_ascii_uppercase();
        let allowed_methods = policy
            .allowed_http_methods
            .iter()
            .map(|item| item.to_ascii_uppercase())
            .collect::<Vec<_>>();
        let denied_methods = policy
            .denied_http_methods
            .iter()
            .map(|item| item.to_ascii_uppercase())
            .collect::<Vec<_>>();
        if denied_methods.contains(&method) {
            deny!("HTTP method explicitly denied");
        }
        if !allowed_methods.is_empty()
            && !allowed_methods.contains(&method)
            && !allowed_methods.iter().any(|item| item == "*")
        {
            deny!("HTTP method not allowed");
        }
        let Some(path) = request.http_path.as_ref() else {
            deny!("HTTP path required for brokered_http");
        };
        if policy
            .denied_http_path_prefixes
            .iter()
            .any(|prefix| path.starts_with(prefix))
        {
            deny!("HTTP path explicitly denied");
        }
        if !policy.allowed_http_path_prefixes.is_empty()
            && !policy
                .allowed_http_path_prefixes
                .iter()
                .any(|prefix| path.starts_with(prefix))
        {
            deny!("HTTP path not allowed");
        }
        if let (Some(limit), Some(size)) = (
            policy.max_http_request_body_bytes,
            request.http_request_body_bytes,
        ) {
            if size > limit {
                deny!("HTTP request body too large");
            }
        }
        Ok(())
    }
}

fn policy_has_peer_binding(policy: &AccessPolicy) -> bool {
    !policy.allowed_os_uids.is_empty()
        || !policy.allowed_executable_paths.is_empty()
        || !policy.allowed_executable_sha256.is_empty()
}

fn assurance_rank(level: &str) -> i32 {
    match level {
        "A0" => 0,
        "A1" => 1,
        "A2" => 2,
        "A3" => 3,
        "A4" => 4,
        "A5" => 5,
        _ => -1,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn request() -> AccessRequest {
        AccessRequest {
            secret_ref: "secret://default/api".to_string(),
            consumer: "cli".to_string(),
            purpose: "deploy".to_string(),
            delivery_mode: DeliveryMode::TerminalPrint,
            provider_assurance: "A1".to_string(),
            raw_export_requested: true,
            http_host: None,
            http_scheme: None,
            http_method: None,
            http_path: None,
            http_request_body_bytes: None,
            os_uid: None,
            executable_path: None,
            executable_sha256: None,
            mfa_verified: false,
            broker_session_id: None,
            now: Utc::now(),
        }
    }

    #[test]
    fn denies_by_default() {
        assert_eq!(
            PolicyEngine::new(vec![]).evaluate(&request()).decision,
            PolicyDecision::Deny
        );
    }

    #[test]
    fn allows_exact_match() {
        let policy = AccessPolicy {
            policy_id: "p1".to_string(),
            secret_refs: vec!["secret://default/api".to_string()],
            allowed_consumers: vec!["cli".to_string()],
            allowed_purposes: vec!["deploy".to_string()],
            allowed_delivery_modes: vec![DeliveryMode::TerminalPrint],
            exportable: true,
            allow_unbound_plaintext_export: true,
            ..serde_json::from_value(serde_json::json!({"policy_id":"defaults"})).unwrap()
        };

        assert!(PolicyEngine::new(vec![policy])
            .evaluate(&request())
            .allowed());
    }

    #[test]
    fn raw_plaintext_export_requires_peer_binding_by_default() {
        let policy = AccessPolicy {
            policy_id: "p1".to_string(),
            secret_refs: vec!["secret://default/api".to_string()],
            allowed_consumers: vec!["cli".to_string()],
            allowed_purposes: vec!["deploy".to_string()],
            allowed_delivery_modes: vec![DeliveryMode::TerminalPrint],
            exportable: true,
            ..serde_json::from_value(serde_json::json!({"policy_id":"defaults"})).unwrap()
        };
        let engine = PolicyEngine::new(vec![policy.clone()]);
        assert_eq!(engine.evaluate(&request()).decision, PolicyDecision::Deny);

        let mut bound_policy = policy;
        bound_policy.allowed_os_uids = vec![1000];
        let mut bound_request = request();
        bound_request.os_uid = Some(1000);
        assert!(PolicyEngine::new(vec![bound_policy])
            .evaluate(&bound_request)
            .allowed());
    }

    #[test]
    fn require_mfa_denies_until_request_is_verified() {
        let policy = AccessPolicy {
            policy_id: "p1".to_string(),
            secret_refs: vec!["secret://default/api".to_string()],
            allowed_consumers: vec!["cli".to_string()],
            allowed_purposes: vec!["deploy".to_string()],
            allowed_delivery_modes: vec![DeliveryMode::TerminalPrint],
            exportable: true,
            allow_unbound_plaintext_export: true,
            require_mfa: true,
            ..serde_json::from_value(serde_json::json!({"policy_id":"defaults"})).unwrap()
        };
        let engine = PolicyEngine::new(vec![policy]);
        let mut request = request();

        let denied = engine.evaluate(&request);
        assert_eq!(denied.decision, PolicyDecision::Deny);

        request.mfa_verified = true;
        assert!(engine.evaluate(&request).allowed());
    }
}
